zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,341 @@
+//! Sapling note and value commitments and types.
+
+#![allow(clippy::unit_arg)]
+
+#[cfg(test)]
+mod arbitrary;
+#[cfg(test)]
+mod test_vectors;
+
+use std::{fmt, io};
+
+use bitvec::prelude::*;
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{
+    keys::sapling::{find_group_hash, Diversifier, TransmissionKey},
+    serde_helpers,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+    types::amount::{Amount, NonNegative},
+};
+
+/// Generates a random scalar from the scalar field 𝔽_{r_𝕁}.
+///
+/// The prime order subgroup 𝕁^(r) is the order-r_𝕁 subgroup of 𝕁 that consists
+/// of the points whose order divides r. This function is useful when generating
+/// the uniform distribution on 𝔽_{r_𝕁} needed for Sapling commitment schemes'
+/// trapdoor generators.
+///
+/// https://zips.z.cash/protocol/protocol.pdf#jubjub
+pub fn generate_trapdoor<T>(csprng: &mut T) -> jubjub::Fr
+where
+    T: RngCore + CryptoRng,
+{
+    let mut bytes = [0u8; 64];
+    csprng.fill_bytes(&mut bytes);
+    // Fr::from_bytes_wide() reduces the input modulo r via Fr::from_u512()
+    jubjub::Fr::from_bytes_wide(&bytes)
+}
+
+/// "...an algebraic hash function with collision resistance (for fixed input
+/// length) derived from assumed hardness of the Discrete Logarithm Problem on
+/// the Jubjub curve."
+///
+/// PedersenHash is used in the definitions of Pedersen commitments (§
+/// 5.4.7.2‘Windowed Pedersen commitments’), and of the Pedersen hash for the
+/// Sapling incremental Merkle tree (§ 5.4.1.3 ‘MerkleCRH^Sapling Hash
+/// Function’).
+///
+/// https://zips.z.cash/protocol/protocol.pdf#concretepedersenhash
+#[allow(non_snake_case)]
+pub fn pedersen_hash_to_point(domain: [u8; 8], M: &BitVec<Lsb0, u8>) -> jubjub::ExtendedPoint {
+    // Expects i to be 0-indexed
+    fn I_i(domain: [u8; 8], i: u32) -> jubjub::ExtendedPoint {
+        find_group_hash(domain, &i.to_le_bytes())
+    }
+
+    /// ⟨Mᵢ⟩
+    ///
+    /// Σ j={0,k-1}: (1 - 2x₂)⋅(1 + x₀ + 2x₁)⋅2^(4⋅j)
+    // XXX: Are internal functions doc'd?
+    //
+    // This is less efficient than it could be so that it can match the math
+    // closely.
+    fn M_i(segment: &BitSlice<Lsb0, u8>) -> jubjub::Fr {
```

Comment:
Suggestion: rather than having three functions defined inside another function, I think it would be better to create a module to contain the helper functions and the function that uses them.  Then the function that uses them can be re-exported for wider use, and all of the helper functions can be documented.